utils/s3/client.cc

```diff
@@ -854,6 +858,167 @@ data_sink client::make_upload_jumbo_sink(sstring object_name, std::optional<unsi
     return data_sink(std::make_unique<upload_jumbo_sink>(shared_from_this(), std::move(object_name), max_parts_per_piece));
 }
 
+// a sink with known object size
+//
+// unlike upload_sink and upload_jumbo_sink, upload_ranged_sink sends the
+// received buffers right away without accumulating them first. if the
+// buffer fed by the caller crosses the boundary of the current part, a
+// new part is created, and the remainder of the buffer is sent to it instead,
+// until whole buffer is consumed.
+class client::upload_ranged_sink final : public upload_sink_base {
+    // TODO: resume support
+    const size_t _part_size;
+    const size_t _total_size;
+    std::optional<future<>> _uploading;
+
+    // avoid stuffing the memory with yet-uploaded buffers in the expense of
+    // having ping-pong between reader (from disk) and writer (to the wire)
+    seastar::pipe<temporary_buffer<char>> _part_body{16};
+
+    future<> upload_part(size_t total_offset, size_t part_size) {
+        // upload a part in a multipart upload, see
+        // https://docs.aws.amazon.com/AmazonS3/latest/API/API_UploadPart.html
+        unsigned part_number = _part_etags.size();
+        _part_etags.emplace_back();
+        auto req = http::request::make("PUT", _client->_host, _object_name);
+        assert(total_offset + part_size <= _total_size);
+        if (total_offset + part_size == _total_size) {
+            assert(part_size > 0);
+        } else {
+            assert(part_size >= aws_minimum_part_size);
+        }
+        req._headers["Content-Length"] = to_sstring(part_size);
+        req.query_parameters.emplace("partNumber", to_sstring(part_number + 1));
+        req.query_parameters.emplace("uploadId", _upload_id);
+        s3l.trace("PUT part {} {}~{}/{} bytes (upload id {})",
+                  part_number, total_offset, part_size, _total_size, _upload_id);
+        req.write_body("bin", part_size, [part_size, this] (output_stream<char>&& out_) mutable -> future<> {
```

Comment:
One problem that we'll face for sure. We know that S3 server likes to close its connections :) For short requests there's https://github.com/scylladb/seastar/pull/1883 that would just replay the request and that's it. And existing sink and jumbo-sink will also be safe in this sense.

This sink is going to be very unstable -- should S3 server close at least _one_ connection and the whole upload needs to be restarted from the beginning.

Reply:
After more thinking on it. Alternatively we can apply this logic:

* If file_size <= 5Gb, then use `upload_sink`. It will create a file just as ranged sink does, with 5Mb parts
* Patch `upload_sink` to have dynamic part size and extend the above up to, say, 10Gb files with 10Mb parts. For backup use-case that's going to be fine
* If file_size > 5Gb (or 10Gb), then use `jumbo_sink`. It will go some slower route with copy-upload

In either case, with scylladb/seastar#1883 connection drops will be replayed just fine

Reply:
Another option -- pass an input_stream<> directory into the ranged sink, so that write_body() callback would read the necessary piece of information from it. Restarting a request would then be natural -- just re-read the data from stream

Reply:
although i don't really understand the "another option", looks like it's promising. =)

by passing an `input_stream<>` directly (i think it was a typo) into the ranged sink, do you mean, ranged sink can read from disk directly without being fed by the caller before it can catch up with the write pointer if current part fail to upload?

Reply:
Yes, that's exactly my though. Otherwise I see no good ways of re-uploading a part if it fails somewhere in the middle of writing the part's body :(